### PR TITLE
RXR-1: tests for as.num and parse_arg plus parse_arg refactor

### PR DESCRIPTION
**Pull request by jasmineirx**
The goal of this ticket was to improve test coverage for json2test, which is sorely lacking. I decided to start with the easiest functions.

I noticed when trying to write examples of unusual inputs for `parse_args` (ex: NULL, or invalid json) that all the code after parsing valid json fails  (i.e., line 33 and onwards) because `config` is not a recognized object in the json2test NAMESPACE. I think that this is code copy/pasted from {opencpu} and thus I _think_ config is supposed to be [this object](https://github.com/opencpu/opencpu/blob/94ac0046e5e30ffae1dce8676817d73cbe443803/R/config.R) created in opencpu but obviously not available in json2test.

We call `parse_arg` in two spots in this package, both to read json files and convert to a list. This function is exported, but I can't think of other places in our code base where we call it. I therefore think changing this function to return only list objects if the character string is valid JSON is appropriate, and so I removed all the code related to returning objects of type expression.

This change goes a bit outside the letter of the ticket, but perhaps not outside the spirit of the ticket (which is to bring json2test up to snuff) and my unit tests should be considerably more expansive if we choose to refactor this function differently (ex: adding opencpu:::config to namespace).

I tested this refactor by ensuring that tests still pass in packages that use json2test for testing (and that if they should fail, they also do fail).

### CODE REVIEW

File: tests/as_num.R [hunk 1]
```diff
@@ -0,0 +1,15 @@
+library(testit)
+
+assert("Parse characters as number - class", is.numeric(json2test:::as.num("7")))
```
Comment (karawoo) [inline tests/as_num.R:3]: I'm not very familiar with testit, does it require the namespace prefix for testing internal functions?

**Review by karawoo — CHANGES_REQUESTED** [review]
Removing the extra stuff from opencpu's `parse_arg()` makes sense to me, it doesn't seem like we need it and I don't think it's possible (certainly not advisable) to import unexported functions from dependencies. So removing the stuff that relies on `opencpu:::config` seems good.

For thorough unit testing I think we should add a test for the error raised if the JSON is invalid. ~I know how to do that in testthat but is there a way to expect errors in testit?~ I guess it would be something like this? Maybe there's a less awkward way...

```r
# missing trailing }
input3 <- "{\"app_type\":\"patient_platform\",\"type\":\"advice\",\"module\":\"vancomycin_adults\",\"isEMR\":null,\"test_api\":true\n"
result <- try(parse_arg(input3), silent = TRUE)
assert("parse_arg throws error for invalid JSON", inherits(result, "try-error"))
```

The Jenkins build looks like it's failing to install rematch2. I see a message that it `new_tibble()` (which rematch2 imports from tibble) is not an exported function from tibble. `new_tibble()` has been exported by tibble for at least 4 years...is it possible the jenkins machine has a tibble version that's that old?

**Review by roninsightrx — APPROVED** [review]
lgtm

File: tests/as_num.R (hunk 1 above)
Comment (roninsightrx) [reply tests/as_num.R:3]: i believe it does, but not 100% sure.

Comment (jasmineirx) [reply tests/as_num.R]: @karawoo, @roninsightrx: I went down a small rabbit hole here. It actually does _not_ require triple colons [if you test through](https://www.rdocumentation.org/packages/testit/versions/0.12) `testit::test_pk()`. However, I did not have the testing structure set up to test in this format, so I have moved around some of our tests.

In the process, I realized I was actually using json2test::assert and not testit::assert. IMO this naming design choice is suboptimal, and should be addressed in a later PR.

See the new commit (which is a substantial number of LOC... apologies!).

But we can now write tests in a way much mroe similar to {testthat} and check tests with `test_pkg("json2test")` (or with `devtools::check()`)

**Comment by jasmineirx** [comment]
@karawoo I implemented a check for the error for invalid json using testit's `has_error` function. It works exactly as your code example, essentially:

```
has_error = function(expr, silent = !interactive()) {
  inherits(try(force(expr), silent = silent), 'try-error')
}
```
(From: [https://github.com/yihui/testit/blob/master/R/testit.R](https://github.com/yihui/testit/blob/master/R/testit.R))

**Review by karawoo — COMMENTED** [review]
This looks good but I don't really understand why the jenkins build is still failing. The error message (`unused argument (silent = TRUE)`) suggests it might still be using `json2test::has_error()` instead of `test::has_error()` but that doesn't make sense 😕

**Comment by jasmineirx** [comment]
@karawoo I figured out why it wasn't running. The `silent` argument to `testit::has_error` was added after version 0.7, which is the version that was installed. I'm not exactly sure why installing a newer version wasn't loading it, so I actually now remove testit and then install it in the Jenkinsfile. It's not an ideal solution, but since we plan to update these containers within the next month it didn't seem worth troubleshooting for now.

**Review by karawoo — APPROVED** [review]
Weird but glad it's working now!

**Pull request merged**